### PR TITLE
Add `dask.array` SVG to the HTML Repr

### PR DESCRIPTION
**Pull request by freyam**
(#7847)

- [ ] Tests added / passed
- [x] Passes `black dask` / `flake8 dask` / `isort dask`

# `Task 2A`

## New Features

- #### Added `chunks` to the `collection_annotations`
- #### Added `collection_annotations` to  `MaterializedLayer` class
- #### Added the Dask Array SVG of the chunks to the HTML Repr
 
## Before

image

## After

```py
import dask.array as da

XL = da.ones((67552, 365941), chunks=(652, 5792))
XL2 = XL.T.astype("f4")
XC = da.ones((365941, 26), chunks=(365941, 26))
LS = da.linalg.lstsq(XC, XL2)[0]
XC_LS = XC @ LS
XLP = XL2 - XC_LS

XLP.dask
```

image
image
image
image
image
image

The computation looks incredible! :fire:

### CODE REVIEW

**Comment by freyam** [comment]
> Tidy up the info we see in the HTML table. For large arrays with many chunks, there wasn't a lot of use in reporting the fully specified chunks - that turns out to be really long, you can see this in Julia's comment here. So, I suggest when we make the HTML table, first we make a copy of the collection_annotations dictionary (like this), then we pop the chunks key out, and finally we make the HTML table from the copied & altered dictionary. 

I will be now working on this.

**Comment by mrocklin** [comment]
It might be worth looking at dask.array.core.Array._repr_html_ (or whatever
it is called).  Some decisions were made there about how to render dask
arrays nicely (including dealing with large chunks attributes).  It might
make sense to reuse that code.

**Comment by freyam** [comment]
Oh yes! I will have a look there and clean up the Output as well

**Comment by GenevieveBuckley** [comment]
Link/copy of comment from the issue thread https://github.com/dask/dask/issues/7847#issuecomment-871891792

> More detail for 2a, above.
> Adding array SVG images to the HTML representation was requested here [#7763 (comment)](https://github.com/dask/dask/pull/7763#issuecomment-858589993)
> 
> > In some more aggressive future we might consider including the dask array SVG array chart directly in the HTML, that's probably future work though
> 
> Suggested approach:
> 
>     * Add `self.chunks` into the `collection_annotations` dictionary. That part of the code is here: https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/array/core.py#L1159
> 
>     * Use the `svg` function to draw the array picture https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/array/svg.py#L7
> 
>     * Add the array picture into the `_repr_html_` method on the `Layer` class https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/highlevelgraph.py#L474
> 
>     * Tidy up the info we see in the HTML table. For large arrays with many chunks, there wasn't a lot of use in reporting the fully specified chunks - that turns out to be really long, you can see this in Julia's comment [here](https://github.com/dask/dask/pull/7763#issuecomment-858658497). So, I suggest when we make the HTML table, first we make a copy of the `collection_annotations` dictionary ([like this](https://www.programiz.com/python-programming/methods/dictionary/copy)), then we pop the chunks key out, and finally we make the HTML table from the copied & altered dictionary. That part of the code is here:
>       https://github.com/dask/dask/blob/e79a56b12512529a2de74122a59d5f69ecc85c6e/dask/highlevelgraph.py#L528

**Comment by GenevieveBuckley** [comment]
> It might be worth looking at dask.array.core.Array._repr_html_ (or whatever it is called). Some decisions were made there about how to render dask arrays nicely (including dealing with large chunks attributes). It might make sense to reuse that code.

I think this code more or less calls the `svg` function directly.

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:502]: Very likely we want this import to live somewhere else before merging (I know you had some problems with the imports earlier)

**Review by GenevieveBuckley — COMMENTED** [review]
A good start!

I'd like to see you:
1. Investigate why you think `dot` is behaving differently, and write that up in a comment (or gist with a notebook, whatever works best)
2. Tidy up the info we see in the HTML table. Earlier we talked about how it starts to look silly if there are lots and lots of chunks shown in there. I gave you some tips earlier for that part (you might even be working on this bit right now :smile: )
3. Start work on 2b whenever you want - it doesn't need to wait until this PR is merged or anything like that.

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py:508]: :heavy_check_mark:

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py:506]: :heavy_check_mark:

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py:502]: I tried several things. No elegant solution was found.

Turns out, it's a case of Circular Imports.
Trying to fix it rn. 

`dask.array` imports `dask.delayed` which imports `HighLevelGraph` which in turn imports `dask.array`

```
ImportError                               Traceback (most recent call last)
<ipython-input-1-dc46e0049dfc> in <module>
----> 1 import dask.array as da
      2 
      3 x = da.random.random((10000, 10000), chunks=(100, 100))
      4 y = x + x.T - x.mean(axis=0)
      5 y.dask

~/GitHub/dask/dask/__init__.py in <module>
      3 from .base import annotate, compute, is_dask_collection, optimize, persist, visualize
      4 from .core import istask
----> 5 from .delayed import delayed
      6 from .local import get_sync as get
      7 

~/GitHub/dask/dask/delayed.py in <module>
     18 from .context import globalmethod
     19 from .core import quote
---> 20 from .highlevelgraph import HighLevelGraph
     21 from .optimization import cull
     22 from .utils import OperatorMethodMixin, apply, ensure_dict, funcname, methodcaller

~/GitHub/dask/dask/highlevelgraph.py in <module>
     23 from .utils import _deprecated, ensure_dict, key_split, stringify
     24 from .utils_test import add, inc  # noqa: F401
---> 25 from dask.array.svg import svg
```

I believe, that's why importing `svg` inside the method works for now as it has no conflicts.

Comment (GenevieveBuckley) [reply dask/highlevelgraph.py:502]: Potentially we could move the svg.py file up, so it doesn't live under the array module anymore. Or keep this solution, delaying the import until inside the function (although I think it should probably have the relative import path with dots, instead of starting with dask dot ...)

We should think about this some more, and then try and work out what the cleanest solution is.

Comment (mrocklin) [reply dask/highlevelgraph.py:502]: In the future I would like to use classes for this.  I would expect array layers to be of some type, like `ArrayLayer`, and then I would expect to call `layer.get_fancy_repr()` or something similar and then the array layers can call the array code dataframe layers can call the dataframe code, etc..

This isn't built out yet though, so we'll probably have to be sloppy until then.

Comment (freyam) [reply dask/highlevelgraph.py:502]: Just for testing, I moved `svg.py` from `dask/array/svg.py` to `dask/svg.py` and then imported that using
```py
from .svg import svg
```
and it works (on top) :tada:. If we plan on using this fix, I will have to change a little code in some files too to make sure the right file is getting imported!

------

Also, I have edited the current absolute addressing to relative and it still works!
```py
from .array.svg import svg
```
works **inside the method**.

**Comment by freyam** [comment]
## Updates

- added a little padding just before the array SVG to make it look clean.
- removed `chunks` from the HTML repr. It's still stored in the `collection_annotations` dictionary if we ever need to use it later for some other purpose.
- investigated why `.dot()` might be interfering with the HTML repr. Turns out it works in some cases and doesn't work in others. I think I am close to figuring out the cause of the issue here.


## Current Demo

### Code

```py
import dask.array as da

x = da.random.random((1000, 1000), chunks=(100, 100))
y = x + x.T - x.mean(axis=0)
y.dask
```

### HTML Representation

image

image

image

image

image


It looks beautiful :heart:

**Comment by freyam** [comment]
@GenevieveBuckley 

Found the error -> There was no error in the first place. I was working with incorrect examples! 
I was taking the dot product of two 1D arrays (whose product is a scalar)!

## A working example
image


## An incorrect example
image

So, I think that solves the issue. I have tried around with other examples too and the HTML gets rendered correctly!

**Comment by freyam** [comment]
Just the one error left

`FAILED dask/tests/test_highgraph.py::test_repr_html_hlg_layers`

```py
from dask.highlevelgraph import HighLevelGraph
hg = HighLevelGraph(
        {"a": {"a": 1, ("a", 0): 2, "b": 3}, "b": {"c": 4}},
        {"a": set(), "b": set()},
    )
hg
```

Error

```
~/GitHub/dask/dask/highlevelgraph.py in _repr_html_(self, layer_index, highlevelgraph_key)
    503 
    504         svg_repr = ""
--> 505         if self.collection_annotations.get("type") == "dask.array.core.Array":
    506             chunks = self.collection_annotations["chunks"]
    507             svg_repr = "<br />" + svg(chunks)

AttributeError: 'NoneType' object has no attribute 'get'
```

**Comment by freyam** [comment]
Unrelated failed tests. I will retrigger once the CI is fixed.

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [reply dask/highlevelgraph.py:502]: If we go with the import happening inside the method here, then I think it makes most sense to put it in this block:

```python
            if chunks is not None:
                from .array.svg import svg
                svg_repr = "<br />" + svg(chunks)
```

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:540]: Please add a comment explaining why we do this

Comment (GenevieveBuckley) [reply dask/highlevelgraph.py:540]: Also, this isn't super extendable if we add other keys to the dictionary that we also want to skip. If we did that, you might want a list of key names to skip, and then check `if key is in keys_to_skip: ...`
Not sure if we need to do that here (probably not, but you need to understand how that would happen if we needed it).

**Review by GenevieveBuckley — CHANGES_REQUESTED** [review]
I like this.
1. We'll need to resolve what's happening with the svg import before this is merged (see discussion thread)
2. I don't think the CI test failures are because of this PR. Can you re-trigger them?
3. The examples look beautiful. Could you replace the "after" screenshots in your top level PR description with some new ones (now that there's been a couple of changes, eg: chunks in the table)

**Comment by jsignell** [comment]
> I don't think the CI test failures are because of this PR. Can you re-trigger them?

There are some known failures on the 3.7 builds (#7907) just ignore those for now.

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py]: 1. Added a comment. 
2. yes, you are right. It would be a lot cleaner if I had a list of keys to not include in the HTML. If we end up having more than 1, i will use a list instead as suggested :heavy_check_mark:

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py]: 1. Moved the import inside, so it only gets called when needed.
2. Surpringly, it's taking the same time as the current Dask implementation.

|                                   dask/dask                                   |                                  freyam/dask                                  |
|:-----------------------------------------------------------------------------:|:-----------------------------------------------------------------------------:|
| 33.7 ns ± 0.743 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each) | 34.3 ns ± 0.615 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each) |


Calculated locally using `%timeit` on the following code.

```py
import dask.array as da

XL = da.ones((67552, 365941), chunks=(652, 5792))
XL2 = XL.T.astype("f4")
XC = da.ones((365941, 26), chunks=(365941, 26))
LS = da.linalg.lstsq(XC, XL2)[0]
XC_LS = XC @ LS
XLP = XL2 - XC_LS

XLP.dask
```

**Comment by freyam** [comment]
## Updates

- [x] Added `chunks` to the `collection_annotations`
- [x] Added `collection_annotations` to  `MaterializedLayer` class
- [x] Added the Dask Array SVG of the chunks to the HTML Repr
- [ ] Figure out how to import `svg` function globally in the  `highlevelgraph.py` file

If anyone has any other suggestions, do let me know.

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [reply dask/highlevelgraph.py]: You'll want timeit only on the very last line `XLP.dask`

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:541]: Unfinished sentence. Suggest rephrasing to a shorter comment. Maybe also placing the comment one line before the if statement begins (that would be two lines, rather than 3)

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:555]: This materialized layer stuff belongs in a separate PR, imo.
Also, isn't there some extra logic about what to do if no `collection_annotations` are passed in (I seem to remember we looked in the dask config, but maybe that was only for `annotations` instead)

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py]: Yes, I had done that.

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py:541]: Oh! fixed it. :heavy_check_mark:

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py:554]: The issue with that is that I need all the layers (`Layers` and `MaterializedLayers` both) to have a `collection_annotations` dictionary.

I guess, for now, I can just add an extra if statement up top and remove the changes I made to `MaterializedLayer` class.

```py
svg_repr = ""
if (
    self.collection_annotations is not None
    and self.collection_annotations.get("type") == "dask.array.core.Array"
):
    chunks = self.collection_annotations.get("chunks")
    if chunks:
        from .array.svg import svg

        svg_repr = "<br />" + svg(chunks)
```

This also works :heavy_check_mark: and I feel it's a better method for now. I will commit this :+1: 

According to my knowledge, the config method doesn't seem to do anything for `annotations` and `collection_annotations` as we only have the following keys only.

```py
config = {
            "temporary-directory": None,
            "dataframe": {"shuffle-compression": None},
            "array": {
                "svg": {"size": 120},
                "slicing": {"split-large-chunks": None},
                "chunk-size": "128MiB",
                "rechunk-threshold": 4,
            },
            "optimization": {
                "fuse": {
                    "active": None,
                    "ave-width": 1,
                    "max-width": None,
                    "max-height": inf,
                    "max-depth-new-edges": None,
                    "subgraphs": None,
                    "rename-keys": True,
                }
            },
        }
   ```

Am I missing out on something? I checked around in the `config,py` as well.

**Comment by freyam** [comment]
## Updates

Just cleaned the code.

@GenevieveBuckley About importing the `svg` function, I think this is the only possible way to do this (keeping in mind that we don't change the file structure or bring the `svg.py` file up a level).

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:537]:
```suggestion
                # Hide verbose chunk details from the HTML table
```

Comment (GenevieveBuckley) [reply dask/highlevelgraph.py:537]: We want to explain the reason *why*, not just what the code is doing. (Reason being, the chunks information is often too verbose to be easily displayed in the HTML table)

**Review by GenevieveBuckley — APPROVED** [review]

**Comment by GenevieveBuckley** [comment]
ping @dask/maintenance 

I think we're basically done here

File: dask/highlevelgraph.py
Comment (freyam) [reply dask/highlevelgraph.py]: Yes, absolutely! Your suggestion is more appropriate and might be useful in the future if we decide to hide more properties.

File: dask/highlevelgraph.py
Comment (GenevieveBuckley) [inline dask/highlevelgraph.py:504]: We talked about this import placement previously, and I think this is probably the best way to leave it

Comment (freyam) [reply dask/highlevelgraph.py:504]: :heavy_check_mark:

**Comment by freyam** [comment]
:yellow_heart:

**Pull request merged**
